Fix crash when targetCoordinate not attached

kadirMete opened the pull request:
This line ensures that if the screen changes (such as during navigation, recomposition, or configuration changes) and the targetCoordinates become detached from the layout tree while the animation is still running, the function will exit early. This prevents any attempt to access properties or methods of a detached LayoutCoordinates, which would otherwise cause a crash (IllegalStateException). Thus, it safely avoids layout-related crashes during UI transitions.

kadirMete committed:
Update ShowcaseView.kt

This line ensures that if the screen changes (such as during navigation, recomposition, or configuration changes) and the targetCoordinates become detached from the layout tree while the animation is still running, the function will exit early. This prevents any attempt to access properties or methods of a detached LayoutCoordinates, which would otherwise cause a crash (IllegalStateException). Thus, it safely avoids layout-related crashes during UI transitions.

jocoand commented:
Hi @kadirMete, thanks for bringing this up!
Could you share a quick example or steps to reproduce the crash?
This would really help me verify. Thanks!

kadirMete commented:
@jocoand Hello, thank you for sharing your library. I encountered an issue when trying to navigate to another screen via a click while the transition flag was set to true. It seems that screen items can still be clicked during the animation transition of the ShowcaseView. This actually led to another problem, so I had to add an overlay box to prevent clicks during the transition.

jocoand commented:
Hi @kadirMete , just to make sure I understand, is the main issue that you want is to prevent clicks on items behind the ShowcaseView ?

Also, I’m still not sure, is this the same issue with the crash you mention before . Does adding the `if (!targetCoordinates.isAttached) return` check help solve this issue?

kadirMete commented:
@jocoand  Actually, yes, the main reason is that during the animation, the user makes a click that causes a screen change (transition: true). If the user changes the screen in some way, it causes this error because targetCoordinates is not attached, and the function continues to run. I will send a video.

`
    java.lang.IllegalStateException: LayoutCoordinate operations are only valid when isAttached is true
                                                                                                    	at androidx.compose.ui.internal.InlineClassHelperKt.throwIllegalStateException(InlineClassHelper.kt:26)
                                                                                                    	at androidx.compose.ui.node.NodeCoordinator.getParentLayoutCoordinates(NodeCoordinator.kt:1673)
                                                                                                    	at androidx.compose.ui.layout.LayoutCoordinatesKt.findRootCoordinates(LayoutCoordinates.kt:280)
                                                                                                    	at androidx.compose.ui.layout.LayoutCoordinatesKt.boundsInRoot(LayoutCoordinates.kt:212)
                                                                                                    	at com.makersoft.uyaniktvmobilelib.presentation.common.showcase.showcaseview.highlight.ShowcaseHighlight$Rectangular.create(ShowcaseHighlight.kt:35)
                                                                                                    	at com.makersoft.uyaniktvmobilelib.presentation.common.showcase.showcaseview.ShowcaseViewKt.ShowcaseView-8B8PKFY(ShowcaseView.kt:70)
                                                                                                    	at com.makersoft.uyaniktvmobilelib.presentation.common.showcase.showcaseview.ShowcaseViewKt.ShowcaseView_8B8PKFY$lambda$7(Unknown Source:28)
                                                                                                    	at com.makersoft.uyaniktvmobilelib.presentation.common.showcase.showcaseview.ShowcaseViewKt.$r8$lambda$yIyEpxkxDfIY7QDROaoJoUO-LQs(Unknown Source:0)
                                                                                                    	at com.makersoft.uyaniktvmobilelib.presentation.common.showcase.showcaseview.ShowcaseViewKt$$ExternalSyntheticLambda2.invoke(D8$$SyntheticClass:0)`

kadirMete commented:
After adding the prevent-click overlay box, I'm not sure if this line is needed, but for safety, I will not remove this line.

video: 
https://drive.google.com/file/d/1Mcdjx6G4fFD3j4Ga36qtaKnj52OJtACq/view?usp=sharing

jocoand commented:
@kadirMete thanks for the videos.

Need to confirm, In the last iteration (where the player launched successfully), after the showcase is dismissed, did you click the `Now turk`  multiple times`?

The first couple click was prevent by the overlay-box.
Then, the animation finished.
Then, the next click will launched the player.

Is the above correct?

kadirMete commented:
@jocoand 

I temporarily removed the overlay I added to record this video. When an overlay is present, screen elements cannot be clicked during the showcase visible or animation but I think still safe add this line.

Actually the crash in the video happens during the animation. in the second part the player opens before the animation is complete.

in both the first and second part of the video, I click during the animation of the next showcase view.

kadirMete commented:
> @jocoand
> 
> I temporarily removed the overlay I added to record this video. When an overlay is present, screen elements cannot be clicked during the showcase visible or animation but I think still safe add this line.
> 
> Actually the crash in the video happens during the animation. in the second part the player opens before the animation is complete.

By the way, in order to prevent the overlay click events I mentioned, I made a change in the overlay I added as in the following code.


`@Composable
fun SequenceShowcase(
    state: SequenceShowcaseState = rememberSequenceShowcaseState(),
    content: @Composable SequenceShowcaseScope.() -> Unit
) {
    val scope = remember(state) { SequenceShowcaseScope(state) }


     //--------------I added to this overlay for prevent click events during animation transition---------------
    if (state.isTransitioning && !state.showCaseVisible) {
        LaunchedEffect(state.isTransitioning) {
            delay(2000)
            state.isTransitioning = false
        }

        Box(
            modifier = Modifier
                .noRippleClickable {}
                .zIndex(5F)
                .fillMaxSize()
        )
    }

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .zIndex(4F)
    ) {
        scope.content()

        state.currentTarget?.let { target ->
            ShowcaseView(
                visible = state.showCaseVisible,
                targetCoordinates = target.coordinates,
                position = target.position,
                alignment = target.alignment,
                highlight = target.highlight,
                animationDuration = target.duration,
                backgroundAlpha = target.backgroundAlpha,
                onFinish = { state.dismiss() },
                onDisplayStateChanged = { displayState ->
                    when (displayState) {
                        ShowcaseDisplayState.Appeared -> {
                            state.onShowcaseViewAppear()
                        }

                        ShowcaseDisplayState.Disappeared -> {
                            state.onShowcaseViewDisappear()
                        }
                    }
                }
            ) { targetRect ->
                target.content(targetRect)
            }
        }
    }
}`

jocoand reviewed (approved):
I think this will a good addition, thanks for your contribution @kadirMete 👍

kadirMete commented:
@jocoand 
You are welcome and thank you for this great package.
